framework/src/media/HttpStream.cpp

```diff
@@ -0,0 +1,183 @@
+namespace stream {
+
+#define SET_OPTION(curl, opt, val) \
+	do {\
+		auto result = curl_easy_setopt(curl, opt, val);\
+		if (result != CURLE_OK) {\
+			meddbg("curl_easy_setopt(%d...) failed!\n", (int)opt);\
+			return false;\
+		}\
+	} while (0)
+
+int HttpStream::mInitializeCount = 0;
+
+std::shared_ptr<HttpStream> HttpStream::create()
+{
+	std::shared_ptr<HttpStream> stream(new HttpStream());
+	if (!stream->init()) {
+		meddbg("HttpStream instance create failed!\n");
+		return nullptr;
+	}
+
+	return stream;
+}
+
+HttpStream::HttpStream() :
+	mCurl(nullptr), mHttpHeaders(nullptr)
+{
+}
+
+HttpStream::~HttpStream()
+{
+	cleanup();
+}
+
+bool HttpStream::addHeader(const std::string &header) {
+	mHttpHeaders = curl_slist_append(mHttpHeaders, header.c_str());
+	if (!mHttpHeaders) {
+		meddbg("curl_slist_append failed, header %s\n", header.c_str());
+		return false;
+	}
+
+	return true;
+}
+
+bool HttpStream::setHeaderCallback(CallbackFunc callback, void *userdata)
+{
+	SET_OPTION(mCurl, CURLOPT_HEADERFUNCTION, callback);
+	SET_OPTION(mCurl, CURLOPT_HEADERDATA, userdata);
+	return true;
+}
+
+bool HttpStream::setWriteCallback(CallbackFunc callback, void *userdata)
+{
+	SET_OPTION(mCurl, CURLOPT_WRITEFUNCTION, callback);
+	SET_OPTION(mCurl, CURLOPT_WRITEDATA, userdata);
+	return true;
+}
+
+bool HttpStream::setReadCallback(CallbackFunc callback, void *userdata)
+{
+	SET_OPTION(mCurl, CURLOPT_READFUNCTION, callback);
+	SET_OPTION(mCurl, CURLOPT_READDATA, userdata);
+	return true;
+}
+
+bool HttpStream::init()
+{
+	if (mInitializeCount == 0) {
+		if (curl_global_init(CURL_GLOBAL_ALL) != CURLE_OK) {
+			meddbg("curl_global_init failed\n");
+			return false;
+		}
+	}
+	mInitializeCount++;
+	mInitializeFlag = true;
+
+	mCurl = curl_easy_init();
+	if (mCurl == nullptr) {
+		meddbg("HttpStream init failed!\n");
+		return false;
+	}
+
+	return true;
+}
+
+void HttpStream::cleanup()
+{
+	if (mHttpHeaders) {
+		curl_slist_free_all(mHttpHeaders);
+		mHttpHeaders = nullptr;
+	}
+
+	if (mCurl != nullptr) {
+		curl_easy_cleanup(mCurl);
+		mCurl = nullptr;
+	}
+
+	if (mInitializeFlag) {
+		mInitializeCount--;
+		if (mInitializeCount == 0) {
+			curl_global_cleanup();
+		}
+		mInitializeFlag = false;
+	}
+}
+
+bool HttpStream::perform()
+{
+	if (mHttpHeaders) {
+		SET_OPTION(mCurl, CURLOPT_HTTPHEADER, mHttpHeaders);
+	}
+
+	CURLcode result = curl_easy_perform(mCurl);
+	if (result != CURLE_OK) {
+		meddbg("curl_easy_perform failed, result %d - %s\n", result, curl_easy_strerror(result));
+		return false;
+	}
+
+	long response_code = 0;
+	result = curl_easy_getinfo(mCurl, CURLINFO_RESPONSE_CODE, &response_code);
+	if (result != CURLE_OK) {
+		meddbg("Get respones failed! result[%d] response[%ld]\n", result, response_code);
+		return false;
+	}
+
+	return true;
+}
+
+bool HttpStream::download(const std::string &url)
+{
+	SET_OPTION(mCurl, CURLOPT_HTTPGET, 1L);
+
+	SET_OPTION(mCurl, CURLOPT_URL, url.c_str());
+
+	if (!perform()) {
+		meddbg("httpget failed!\n");
```

Comment:
please add space between "https" and  "get"

Reply:
Fixed